### PR TITLE
Added variant for Waveshare ESP32-S3-Zero

### PR DESCRIPTION
**Pull request by Y1hsiaochunnn**
## Description of Change
Added variant for ESP32-S3-Zero with RGB Light

## Tests scenarios
Untested, pin definition is based on information in Waveshare wiki.

## Related links

Wiki For [ESP32-S3-Zero](https://www.waveshare.com/wiki/ESP32-S3-Zero)

Pending USB VID PR: https://github.com/espressif/usb-pids/pull/174

### CODE REVIEW

File: variants/waveshare_esp32_s3_zero/pins_arduino.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+static const uint8_t OUTPUT_IO9 = 9;
+static const uint8_t OUTPUT_IO10 = 10;
+static const uint8_t OUTPUT_IO11 = 11;
+static const uint8_t OUTPUT_IO12 = 12;
+static const uint8_t OUTPUT_IO13 = 13;
+
+// UART0 pins
+static const uint8_t TX = 43;
+static const uint8_t RX = 44;
+
```
Comment (P-R-O-C-H-Y) [inline variants/waveshare_esp32_s3_zero/pins_arduino.h:33]: I2C and SPI pins definitions are missing, so CI is failing. Please add those.
If you don't want to specify the pins, you can use -1 as the PIN number.

**Comment by Y1hsiaochunnn** [comment]
Hi @P-R-O-C-H-Y ,
how does it look now?

File: boards.txt [hunk 2]
```diff
@@ -38600,6 +38600,198 @@ Geekble_ESP32C3.menu.EraseFlash.none.upload.erase_cmd=
...
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB=RainMaker 4MB No OTA
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.build.partitions=rainmaker_4MB_no_ota
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.upload.maximum_size=4038656
+
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs=OTA no FS (2MB APP with OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.build.custom_partitions=partitions_otanofs_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.upload.maximum_size=2031616
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app=Max APP (4MB APP no OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.build.custom_partitions=partitions_all_app_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.upload.maximum_size=4128768
```
Comment (P-R-O-C-H-Y) [inline boards.txt:38743]: I cannot see those partitions anywhere. The one you added is also not listed here.

**Comment by Y1hsiaochunnn** [comment]
Fixed

File: boards.txt [hunk 3]
```diff
@@ -38600,6 +38600,198 @@ Geekble_ESP32C3.menu.EraseFlash.none.upload.erase_cmd=
...
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB=RainMaker 4MB No OTA
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.build.partitions=rainmaker_4MB_no_ota
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.upload.maximum_size=4038656
+
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs=OTA no FS (2MB APP with OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.build.custom_partitions=partitions_otanofs_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.upload.maximum_size=2031616
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app=Max APP (4MB APP no OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.build.custom_partitions=partitions_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.upload.maximum_size=4128768
```
Comment (P-R-O-C-H-Y) [inline boards.txt:38743]: This partitions says no OTA, but in the partitions there is ota_0 and ota_1.

```
# Name,    Type,  SubType,     Offset,      Size,       Flags
 nvs,       data,  nvs,         0x9000,      0x5000,
 otadata,   data,  ota,         0xe000,      0x2000,
 ota_0,     app,   ota_0,       0x10000,     0x1F0000,
 ota_1,     app,   ota_1,       0x200000,    0x1F0000,
```

So you can delete ota_1 and change the size of ota_0 to match flash size.

Comment (Y1hsiaochunnn) [reply boards.txt:38743]: ok,ill do it

File: variants/waveshare_esp32_s3_zero/partitions_4MB.csv
Comment (P-R-O-C-H-Y) [inline variants/waveshare_esp32_s3_zero/partitions_4MB.csv:1]: This doesn't look good. You can use this partition table as a reference, but change it to fit in 4MB.
Its 8MB (no OTA, no FS)
```
# Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     0x9000,  0x5000,
otadata,  data, ota,     0xe000,  0x2000,
app0,     app,  factory, 0x10000, 0x7E0000,
coredump, data, coredump,0x7F0000,0x10000,

```

**Comment by P-R-O-C-H-Y** [comment]
@Y1hsiaochunnn Also move the partitions from `variant` folder to `tools/partitions`.
Renaming to something like max_app_4MB and ota_nofs_4MB. Thanks

**Comment by Y1hsiaochunnn** [comment]
i did it so,network is not good, so I amend it repeatedly

File: tools/partitions/max_app_4MB.csv [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Name,   Type, SubType, Offset,  Size, Flags
+nvs,      data, nvs,     0x9000,  0x5000,
+otadata,  data, ota,     0xe000,  0x2000,
+app0,     app,  factory, 0x10000, 0x360000,
+coredump, data, coredump,0x7F0000,0x10000,
```
Comment (P-R-O-C-H-Y) [inline tools/partitions/max_app_4MB.csv:5]: Fix partition to use all available memory
```suggestion
# Name,   Type, SubType, Offset,  Size, Flags
nvs,      data, nvs,     0x9000,  0x5000,
otadata,  data, ota,     0xe000,  0x2000,
app0,     app,  factory, 0x10000, 0x3E0000,
coredump, data, coredump,0x3F0000,0x10000,
```

File: boards.txt [hunk 5]
```diff
@@ -38600,6 +38600,198 @@ Geekble_ESP32C3.menu.EraseFlash.none.upload.erase_cmd=
...
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB=RainMaker 4MB No OTA
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.build.partitions=rainmaker_4MB_no_ota
+waveshare_esp32_s3_zero.menu.PartitionScheme.rainmaker_4MB.upload.maximum_size=4038656
+
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs=OTA no FS (2MB APP with OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.build.custom_partitions=ota_nofs_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.otanofs.upload.maximum_size=2031616
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app=Max APP (4MB APP no OTA)
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.build.custom_partitions=max_app_4MB
+waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.upload.maximum_size=4128768
```
Comment (P-R-O-C-H-Y) [inline boards.txt:38743]: Maximum upload size needs to match the app partition size -> 0x3E0000 = 4063232
```suggestion
waveshare_esp32_s3_zero.menu.PartitionScheme.all_app.upload.maximum_size=4063232
```

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@Y1hsiaochunnn PTAL

**Review by P-R-O-C-H-Y — APPROVED** [review]
LGTM :)

**Pull request merged**